dali/operators/image/convolution/laplacian_params.h

```diff
@@ -0,0 +1,276 @@
+  }
+
+  bool HasSmoothingSpecified() {
+    return has_smooth_const_ || has_smooth_tensor_;
+  }
+
+  bool HasScaleSpecified() {
+    return has_scales_const_ || has_scales_tensor_;
+  }
+
+  void ObtainWindowSizes(const ArgumentWorkspace &ws, int nsamples) {
+    int prev_size = window_sizes_.size();
+    if (HasPerSampleWindows() || prev_size < nsamples) {
+      window_sizes_.resize(nsamples);
+    }
+    int sample_idx = HasPerSampleWindows() ? 0 : prev_size;
+    for (; sample_idx < nsamples; sample_idx++) {
+      SetSampleWindowSizes(ws, sample_idx);
+    }
+  }
+
+  void SetSampleWindowSizes(const ArgumentWorkspace &ws, int sample_idx) {
+    auto &window_sizes = window_sizes_[sample_idx];
+    std::array<int, axes> deriv_arg;
+    GetGeneralizedArg<int>(make_span(deriv_arg), windowSizeArgName, sample_idx, spec_, ws);
+    for (auto d_size : deriv_arg) {
+      DALI_ENFORCE(3 <= d_size && d_size <= maxWindowSize && d_size % 2 == 1,
+                   make_string("Window size must be an odd integer between 3 and ", maxWindowSize,
+                               ", got ", d_size, " for sample: ", sample_idx, "."));
+    }
+    if (!HasSmoothingSpecified()) {
+      for (int i = 0; i < axes; i++) {
+        std::fill(window_sizes[i].begin(), window_sizes[i].end(), deriv_arg[i]);
+      }
+    } else {
+      std::array<int, axes> smooth_arg;
+      GetGeneralizedArg<int>(make_span(smooth_arg), smoothingSizeArgName, sample_idx, spec_, ws);
+      for (auto s_size : smooth_arg) {
+        DALI_ENFORCE(
+            1 <= s_size && s_size <= maxWindowSize && s_size % 2 == 1,
+            make_string("Smoothing window size must be an odd integer between 1 and ",
+                        maxWindowSize, ", got ", s_size, " for sample: ", sample_idx, "."));
+      }
+      for (int i = 0; i < axes; i++) {
+        for (int j = 0; j < axes; j++) {
+          window_sizes[i][j] = i == j ? deriv_arg[i] : smooth_arg[j];
+        }
+      }
+    }
+  }
+
+  void ComputeScales(int nsamples) {
+    if (HasPerSampleWindows()) {
+      scales_.resize(nsamples);
+      for (int sample_idx = 0; sample_idx < nsamples; sample_idx++) {
+        scales_[sample_idx] = GetNormalizationFactors<axes>(window_sizes_[sample_idx]);
+      }
+    } else {
+      int prev_size = scales_.size();
+      if (prev_size < nsamples) {
+        auto scale = prev_size == 0 ? GetNormalizationFactors<axes>(window_sizes_[0]) : scales_[0];
+        scales_.resize(nsamples, scale);
+      }
+    }
+  }
+
+  void ObtainScales(const ArgumentWorkspace &ws, int nsamples) {
+    int prev_size = scales_.size();
+    if (has_scales_tensor_ || prev_size < nsamples) {
+      scales_.resize(nsamples);
+    }
+    for (int sample_idx = has_scales_tensor_ ? 0 : prev_size; sample_idx < nsamples; sample_idx++) {
+      GetGeneralizedArg<float>(make_span(scales_[sample_idx]), scaleArgName, sample_idx, spec_, ws);
+    }
+  }
+
+  static constexpr int num_windows = axes * axes;
+  OpSpec spec_;
```

Comment:
I find this unnecessary. Each operator has a `spec_`, and you could pass it to the relevant APIs. We do that with other *Params structs.

Reply:
done